elixir-phoenix-deployement.md

```diff
@@ -0,0 +1,162 @@
+![image](https://user-images.githubusercontent.com/6057298/83642784-2e1d1200-a5a7-11ea-8c97-9c7dd920469c.png)
+
+pricing:
+![image](https://user-images.githubusercontent.com/6057298/83641712-cc0fdd00-a5a5-11ea-8cbf-49981ae3747e.png)
+
+The pricing above doesn't include the database which needs to be added to the total cost, see https://elements.heroku.com/addons/heroku-postgresql
+
+
+### Gigalixir
+
+Similar to Heroku [Gigalixir](https://www.gigalixir.com/) provides
+a platform which focuses _exclusively_ on deploying Elixir/Phoenix application.
+Having an exclusive focus on deploying Elixir Apps gives Gigalixir several key advantages:
+1. Zero-downtime blue-green continuous deployment.
+see: https://martinfowler.com/bliki/BlueGreenDeployment.html
+2. No limit to concurrent connections.
+3. All clustering handled transparently. 
+(so if you need to scale your app beyond 10k concurrent users, 
+you don't have to pay exponentially more for bigger "dynos" 
+the way you are forced to scale vertically on Heroku)
+
+Deploying: https://elixircasts.io/deploying-with-gigalixir-%28revised%29
+
+Pricing:
+
+![image](https://user-images.githubusercontent.com/6057298/83643629-290c9280-a5a8-11ea-89da-648daec204f3.png)
+
+Gigalixir vs Heroku:
+
+![image](https://user-images.githubusercontent.com/6057298/83644108-d1225b80-a5a8-11ea-869a-c8c0e4a28012.png)
+
+Like Heroku you need to add the cost for using the database:
+
+![image](https://user-images.githubusercontent.com/6057298/83644296-09c23500-a5a9-11ea-9fdd-c50532195c95.png)
+
+see also tiers pricing page: https://gigalixir.readthedocs.io/en/latest/tiers-pricing.html
+
+
+### Render
+
+[Render](https://render.com/) is another Paas similar to Heroku
+
+pricing:
+
+![image](https://user-images.githubusercontent.com/6057298/83645052-e9df4100-a5a9-11ea-8d3e-f1b18bb4cc02.png)
+
+
+
+## Virtual Private Server
+
+VPS allow us to manage ourself the deployement
+setup. This allow us to customise the server and
+the costs linked to it.
+
+
+### Linode
+
+#### Ubuntu
+
+Linode provides and support Ubuntu:
+![image](https://user-images.githubusercontent.com/6057298/83647811-3bd59600-a5ad-11ea-893c-b99e3df7f605.png)
+
+
+From there the idea is install Erlang and Elixir on the server
+and then to run the application.
+
+- Install Erlang/Elixir with asdf:
+    - `git clone https://github.com/asdf-vm/asdf.git ~/.asdf --branch v0.7.8`
+    - Edit `~/.bashrc file` and add `. $HOME/.asdf/asdf.sh` and run `source ~/.barhrc` to access the `asdf` command
+    - Install required pacakges for Erlang `sudo apt install libssl-dev make automake autoconf libncurses5-dev gcc`
+    - Add erlang plugin to asdf: `asdf plugin-add erlang`
+    - Install Erlang: `asdf install erlang latest`
+    - Install Elixir: `asdf install elixir latest`
+    - Define which Elixir version to use `asdf global elxir <version>`
+
+- Instsall Nodejs using nvm
```

Comment:
We _definitely_ don't want to install `node.js` or any of the node **build dependencies** on the Production Server. 
This is a _massive_ attack vector that would allow an attacker who successfully compromised one of the  _thousands_ of dependencies in the [`package-lock.json`](https://github.com/dwyl/hits/blob/master/assets/package-lock.json) file to create a `postinstall` script that leaks our AWS, DB or Encryption keys to a remote server. 
My (_security mindset_) reasoning for wanting to use `OpenBSD` on the Production server to _minimise_ the number of attack vectors would be _completely_ negated by installing `node.js` on the VPS. 

I _tolerate_ running Node.js on my `localhost` because I still maintain a bunch of NPM packages,
but I _definitely_ don't want to run Node.js on my Server unless I _have_ to (_e.g. because the Application is written in JS..._)